File: Trees/Binary Tree Level Order Traversal/BinaryTreeLevelOrderTraversal.java
// Use queue to represent nodes at current level. For each level, track current queue/level size,  add node values to the level list and add their non-null children for the next level. Add level list to result.
public class BinaryTreeLevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int queueSize = queue.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < queueSize; i++) {
                TreeNode currNode = queue.poll();
                level.add(currNode.val);
                if (currNode.left != null) queue.offer(currNode.left);
                if (currNode.right != null) queue.offer(currNode.right);
            }
            result.add(level);
        }

        return result;
    }
}

File: Trees/Binary Tree Level Order Traversal/binary_tree_level_order_traversal.cpp
class BinaryTreeLevelOrderTraversal {
public:
    // BFS
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};
        vector<vector<int>> result;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            vector<int> level = {};
            for (int i = 0; i < size; ++i) {
                TreeNode* curr = q.front();
                q.pop();
                level.push_back(curr->val);
                if (curr->left) q.push(curr->left);
                if (curr->right) q.push(curr->right);
            }
            result.push_back(level);
        }

        return result;
    }
};
